nixos/modules/services/monitoring/prometheus/default.nix

```diff
@@ -928,23 +975,67 @@ in {
     systemd.services.prometheus = {
       wantedBy = [ "multi-user.target" ];
       after    = [ "network.target" ];
-      preStart = ''
+      preStart = mkIf (!cfg.enableReload) ''
          ${lib.getBin pkgs.envsubst}/bin/envsubst -o "/run/prometheus/prometheus-substituted.yaml" \
                                                   -i "${prometheusYml}"
       '';
       serviceConfig = {
         ExecStart = "${cfg.package}/bin/prometheus" +
           optionalString (length cmdlineArgs != 0) (" \\\n  " +
             concatStringsSep " \\\n  " cmdlineArgs);
+        ExecReload = mkIf cfg.enableReload "+${reload}/bin/reload-prometheus";
```

Comment:
Without having dug too deep into the internals here, couldn't this just be a combination of `write-prometheus-config` and a subsequent `reload-prometheus`? If I'm understanding this correctly, the other two units below are only for generating and recreating the configuration file. So why can't the main Prometheus service get `reloadIfChanged` and handle that itself – am I missing something here? You mention above that you want `switch-to-configuration` to only succeed when the new configuration is applied, but how does that require it's own unit?

Reply:
The problem is that when the `prometheus` service has `reloadIfChanged=true` it will reload on *all* changes. That would be fine if you only change the config file but if you upgrade the prometheus package it would also just reload instead of the needed restart. That's why I need a separate service.

Reply:
I see, it makes sense in that case. Your solution is probably cleaner than comparing the new executable's version with the current one.